### ISSUE TITLE
Moving through nested menus can cause them to erroniously close

### ISSUE THREAD

**Issue by ZingBallyhoo** [ocornut/imgui#7325]
### Version/Branch of Dear ImGui:

Version 1.90.4 WIP (19031), Branch: master

### Back-ends:

imgui_impl_win32.cpp + imgui_impl_dx11.cpp

### Compiler, OS:

Windows 10 + MSVC

### Full config/build information:

```
Dear ImGui 1.90.4 WIP (19031)
--------------------------------
sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=199711
define: _WIN32
define: _WIN64
define: _MSC_VER=1938
define: _MSVC_LANG=201402
--------------------------------
io.BackendPlatformName: imgui_impl_win32
io.BackendRendererName: imgui_impl_dx11
io.ConfigFlags: 0x00000003
 NavEnableKeyboard
 NavEnableGamepad
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000000E
 HasMouseCursors
 HasSetMousePos
 RendererHasVtxOffset
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1264.00,761.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

### Details:

**My Issue/Question:**

I don't think I can describe this issue better than the title, please see the videos.
The same thing happens with any further level of nesting (always closes up to the same place, in this case "Nested2")

Introduced in commit https://github.com/ocornut/imgui/commit/76e09c4b0faed1c7e12ebb660b667b5abfcc6679 (part of 1.90.2)

### Screenshots/Video:

| Current Behavior | Expected Behavior |
| ------------- | ------------- |
| <video src="https://user-images.githubusercontent.com/13888932/305573157-ea89da54-16f9-4a03-a454-18f20f258899.mp4">  | <video src="https://user-images.githubusercontent.com/13888932/305574514-e752fc74-cc1b-495c-81cf-2c01cc3dbff9.mp4">|

### Minimal, Complete and Verifiable Example code:

```cpp
if (ImGui::Begin("MenuRepro"))
{
    if (ImGui::BeginMenu("Nested1"))
    {
        if (ImGui::BeginMenu("Nested2"))
        {
            for (int i = 0; i < 10; i++)
            {
                if (ImGui::BeginMenu(std::to_string(i).c_str())) ImGui::EndMenu();
            }
            ImGui::EndMenu();
        }
        ImGui::EndMenu();
    }
}
ImGui::End();
```

**Comment by ocornut** [comment]
Thank you for the report. Should be fixed 014e0ac.
Previous commit 76e09c4b is only accidentally responsible that it made another issue surface, but the logic of it seems correct.

When moving from "A" to "B":
- "A" closes itself in BeginMenu(), which focus parent.
- "B" opens (visible and focused next frame)

When moving from "B" to "A":
- "A" opens which triggers a popup closure, otherwise that specific call to `ClosePopupToLevel()` had `restore_focus_to_window_under_popup = false` which here is incorrect (false is only used by FocusXXX functions), so focus was kept on "A" 
- "B" opens (aimed visible and focused next frame)
- NewFrame() detect A is focused but not visible anymore, tries to focus top-most window using `FocusTopMostWindowUnderOne(NULL, NULL, NULL, ImGuiFocusRequestFlags_RestoreFocusedChild);` but because this looks root window only it would revert focus to the root menus, which in turns close all its sub-menus.

So I fixed the call to `ClosePopupToLevel()` to use `restore_focus_to_window_under_popup = true` which seems correct and passes all my tests (and checked where it was introduced, in d31fe97f7).
But it's not impossible that the logic of `FocusTopMostWindowUnderOne()` would be improved in the future, as ideally it should be usable in chain of child menus (I don't have a clear situation in mind where it would be problematic but I can envision it is possible).

I was puzzled by why it didn't happen in the demo, but this requires: 
- two adjacent sibling menus
- moving upward
- not being in the root menu
The only situation where this exists in the demo is with the "Options"/"Colors" menu when browsed from the recursive path, but the demo also show appending to a menu with the "Options" and this accidentally "fixes" the issue.

Added test https://github.com/ocornut/imgui_test_engine/commit/e4254ea32e99e3f8480846bb7a08b43357464a24

**Comment by ZingBallyhoo** [comment]
I am actually still able to reproduce this on 1.90.4.. but only when the mouse is moving fast (skipping over items?)

https://github.com/ocornut/imgui/assets/13888932/a8d0ca37-d09e-4f7e-86bc-5d49b47ad811

**Comment by ocornut** [comment]
I can confirm there is an issue when trying to open a sibling menu two frames in a row. I'm investigating.

**Comment by ocornut** [comment]
I spent quite a few hours investigating what the proper solution was, and in the end I think this simple fix is the most correct thing to do: c3f8f4d

I have amended the test for it : https://github.com/ocornut/imgui_test_engine/commit/a35d13af6c7d9d649158b02c2a619f3fac5d1e27

Thanks again!

Some info, when we close a popup via `ClosePopupToLevel()` we rely on info stored in `g.OpenPopupStack[]` for that popup, but that info is incomplete if the popup window has never been Begin()-ed into.

- Frame N: Mouse is over Item 3
- Frame N+1: Move to item 2: when item 2 detects hover is requests popup open but can't display immediately (technically in this specific ordering it could, but the logic is to avoid recycling same popup host window in a same frame).
- Frame N+2: Move to item 1: when item 1 detects hover it requests popup open, which closes previous one, that closure normally first restore previous focus because opening the new window.
